### ISSUE TITLE
Best ways to handle failure to find a collision-free hash

### ISSUE THREAD

**Issue by sqs** [alecthomas/mph#6]
(I apologize if this is out of scope for this library, and for my relatively shallow understanding of MPHs.)

Using the FNV hash func and even some simple inputs (e.g., keys `c` and `c2`, where `c` is any character), the builder often fails to find a collision-free hash. This is not a limitation of this particular project; implementations of the same algorithm in other languages also frequently fail on this class of inputs. (I say "often" and "frequently" because the random seed is the current time, and sometimes it succeeds.)

Is there a generally accepted solution for handling this occurrence, such as retrying the generation with a different random seed? Or allowing a small number of collisions (and iterating over the collisions in Get) if the cardinality is small? (It might make the most sense to add support for these things in external code, not in this core mph lib. But I notice that gperf supports producing imperfect hashes, so there is some precedent for including them.)

So, basically, how do other people handle the failure to find a collision-free hash in their own code? (Just fall back to a map or another hashing scheme?)

**Comment by sqs** [comment]
In the [CHD paper](http://cmph.sourceforge.net/papers/chm92.pdf) I see that `n/m` determines (at least in part) the probability of finding a collision-free hash:

![image](https://cloud.githubusercontent.com/assets/1976/5867591/11bca948-a253-11e4-9ad4-dfc2151444d0.png)

So the solution might just be to bump up `n` for small keysets. Will investigate.

**Comment by sqs** [comment]
It seems this library chooses `n=len(keys)` and then lets `m=n/c` (where `c=2`). However, the CHD paper's Simple Example (Section 4) chooses `m=len(keys)` (and defines `m` to be "the number of words") and then lets `n=m*c` (where `c~=2.08`). And in general, the CHD paper talks about choosing `m` first then determining `n` later.

What are the repercussions of choosing `m` such that `m << len(keys)`? I can't find any discussion of that in the paper. Was this an intentional choice in this library?

Sorry if I've misunderstood anything!

**Comment by alecthomas** [comment]
The paper you are reading is for [CHM](http://cmph.sourceforge.net/chm.html) (not CHD, which this package's implementation is based on is [CHD](http://cmph.sourceforge.net/chd.html). Take a look at that and see what you think.

**Comment by sqs** [comment]
Whoops, sorry. I'll take a look at the right paper.
